packages/components/tests/dummy/app/controllers/components/table.js

```diff
@@ -148,4 +179,251 @@ export default class ComponentsTableController extends Controller {
     }
     return clonedModelClusters;
   };
+
+  // GENERIC MULTI-SELECT FUNCTIONALITIES
+
+  @action
+  onSelectionChangeLogArguments() {
+    console.log(...arguments);
+  }
+
+  @action
+  mockIndeterminateState(checkbox) {
+    checkbox.indeterminate = true;
+  }
+
+  // MULTI-SELECT DEMO #1
+  // Multi-select table with external filter for odd/even rows
+
+  get multiSelectFilteredData__demo1() {
+    if (this.multiSelectFilterRows__demo1 === 'all') {
+      return this.multiSelectModelData__demo1;
+    } else {
+      const remainder = this.multiSelectFilterRows__demo1 === 'even' ? 0 : 1;
+      return this.multiSelectModelData__demo1.filter(
+        (item) => item.id % 2 === remainder
+      );
+    }
+  }
+
+  @action
+  toggleMultiSelectToggleScope__demo1(event) {
+    this.multiSelectToggleScope__demo1 = event.target.checked;
+  }
+
+  @action
+  toggleMultiSelectToggleDebug__demo1(event) {
+    this.multiSelectToggleDebug__demo1 = event.target.checked;
+  }
+
+  @action
+  onChangeMultiSelectFilter__demo1(event) {
+    this.multiSelectFilterRows__demo1 = event.target.value;
+  }
+
+  @action
+  onSelectionChangeWithModel__demo1({
+    selectionKey,
+    selectionCheckboxElement,
+    selectableRowsStates,
+  }) {
+    console.log(...arguments);
+    if (selectionKey === 'all' && this.multiSelectToggleScope__demo1) {
+      const selectAllState = selectionCheckboxElement.checked;
+      this.multiSelectModelData__demo1.forEach((modelRow) => {
+        modelRow.isSelected = selectAllState;
+      });
+    } else {
+      selectableRowsStates.forEach((row) => {
+        const recordToUpdate = this.multiSelectModelData__demo1.find(
+          (modelRow) => modelRow.id === row.selectionKey
+        );
+        if (recordToUpdate) {
+          recordToUpdate.isSelected = row.isSelected;
+        }
+      });
```

Comment:
I gave some thought to this, in conjunction with the conversation in Slack.

I think it works just fine as it is, but I added a suggestion if we're determined to improve computation.

In the current implementation we have a time complexity of O(n*m) in the worst case (n is the length of `selectableRowsStates`; m is the length of `multiSelectModelData__demo1`). That's because for each row in `selectableRowsStates`, we are searching for a matching row in `multiSelectModelData__demo1` using the `find` method, which can take up to m iteration.

If we create a map of `multiSelectModelData__demo1` by `id`, it will allow us to find the matching row in constant time, reducing the time complexity to O(n+m).

```suggestion
      const modelDataMap = this.multiSelectModelData__demo1.reduce(
        (map, modelRow) => {
          map[modelRow.id] = modelRow;
          return map;
        },
        {}
      );

      selectableRowsStates.forEach((row) => {
        const recordToUpdate = modelDataMap[row.selectionKey];
        if (recordToUpdate) {
          recordToUpdate.isSelected = row.isSelected;
        }
      });
```

Arguably, the proposed solution is not easier to read than what we currently have. And for a reasonable number of rows, it won't make a big difference.

Reply:
If it's more of a best practice to reduce time complexity in this way then it could be useful to update from that regard (to serve as a better example). @didoo what do you think?

Reply:
Addressed here: https://github.com/hashicorp/design-system/pull/1859/commits/573baf9e3de7bd39352fb00b13c4182b84d25e40